src/main/jbake/content/blog/2017/07/opentracing-jaxrs.adoc

```diff
@@ -0,0 +1,136 @@
+custom data to the trace.
+
+== Demo application
+Creating JAX-RS app from scratch might be a time consuming task, therefore in this case we are going to use
+Wildfly Swarm and use http://wildfly-swarm.io/generator/[app generator]! Choose JAX-RS and CDI dependencies and hit generate button.
+
+ifndef::env-github[]
+image::/img/blog/2017/wf-swarm-generator.png[caption="Figure 1: ", title="Wildfly Swarm generator."]
+endif::[]
+ifdef::env-github[]
+image::../../../../../assets/img/blog/2017/wf-swarm-generator.png[caption="Figure 1: ", title="Wildfly Swarm generator."]
+endif::[]
+
+The generated application contains one REST endpoint which returns hello world string.
+This endpoint is accessible on `http://localhost:8080/hello`.
+It is sufficient for our demo and in the next step we are going to add instrumentation and simple business logic.
+Do not worry at the bottom of this article are all the necessary links including source code.
+
+== Instrumentation
+To add JAX-RS OpenTracing instrumentation is very simple, just include the following dependency on
+the classpath and tracing feature will be automatically registered.
+
+----
+<dependency>
+    <groupId>io.opentracing.contrib</groupId>
+    <artifactId>opentracing-jaxrs2</artifactId>
+</dependency>
+----
+
+OpenTracing is just an API, therefore it is required to register a specific tracer instance. In this demo
+we are going to use https://github.com/uber/jaeger[Jaeger tracing system]. Tracer should be created
+and initialized only once per process, hence `ServletContextListener` is the ideal place for this task:
+
+[source,java]
+----
+@WebListener
+public class TracingContextListener implements ServletContextListener {
+
+  @Inject
+  private io.opentracing.Tracer tracer;
+
+  @Override
+  public void contextInitialized(ServletContextEvent sce) {
+    GlobalTracer.register(jaegerServer());
+  }
+
+  @Override
+  public void contextDestroyed(ServletContextEvent sce) {}
+
+  @Produces
+  @Singleton
+  public static io.opentracing.Tracer jaegerTracer() {
+    return new Configuration("wildfly-swarm", new Configuration.SamplerConfiguration(
+        ProbabilisticSampler.TYPE, 1),
+        new Configuration.ReporterConfiguration())
+        .getTracer();
+  }
+}
+----
+
+As you noticed we are using CDI to share a tracer instance in our app. If we forgot to register
+specific tracer instance then tracing feature would use `NoopTracer`. Now we can verify tracing
+by starting Jaeger server `docker run --rm -it --network=host jaegertracing/all-in-one` and
+hitting hello endpoint. Trace with one span should be present in the UI
+accessible on `http://localhost:16686`.
+
+== Instrumenting business logic
+JAX-RS instrumentation provides out of the box a nice visibility into your app, however, often
+it's necessary to add custom data to the trace to see what is happening in service or database layer.
+
+Following code snippet shows how service layer can create and add data to the trace:
```

Comment:
The following code snippet shows how the service layer can create and add data to the trace.